nf_core/modules/update.py

```diff
@@ -390,32 +419,51 @@ def get_all_modules_info(self):
                 f"Overriding '--sha' flag for module{plural_s(overridden_modules)} with "
                 f"'.nf-core.yml' entry: '{overridden_str}'"
             )
+        # Loop through modules_info and create on ModulesRepo object per remote and branch
+        repos_and_branches = {}
+        for repo_name, mods in modules_info.items():
+            for mod, sha, mod_branch in mods:
+                if branch is not None:
+                    mod_branch = branch
+                if (repo_name, mod_branch) not in repos_and_branches:
+                    repos_and_branches[(repo_name, mod_branch)] = []
+                repos_and_branches[(repo_name, mod_branch)].append((mod, sha))
 
         # Get the git urls from the modules.json
-        modules_info = [
-            (self.modules_json.get_git_url(repo_name), self.modules_json.get_base_path(repo_name), mods_shas)
-            for repo_name, mods_shas in modules_info.items()
-        ]
+        modules_info = (
+            (self.modules_json.get_git_url(repo_name), branch, self.modules_json.get_base_path(repo_name), mods_shas)
+            for (repo_name, branch), mods_shas in repos_and_branches.items()
+        )
 
         # Create ModulesRepo objects
-        modules_info = [
-            (ModulesRepo(remote_url=repo_url, base_path=base_path), mods_shas)
-            for repo_url, base_path, mods_shas in modules_info
-        ]
+        repo_objs_mods = []
+        for repo_url, branch, base_path, mods_shas in modules_info:
+            try:
+                modules_repo = ModulesRepo(remote_url=repo_url, branch=branch, base_path=base_path)
+            except LookupError as e:
+                log.warning(e)
+                log.info(f"Skipping modules in '{modules_repo.fullname}'")
+            else:
+                repo_objs_mods.append((modules_repo, mods_shas))
 
-        # Flatten and return the list
-        modules_info = [(repo, mod, sha) for repo, mods_shas in modules_info for mod, sha in mods_shas]
+        # Flatten the list
+        modules_info = [(repo, mod, sha) for repo, mods_shas in repo_objs_mods for mod, sha in mods_shas]
 
-        # Verify that that all modules exist in their respective ModulesRepo,
+        # Verify that that all modules and shas exist in their respective ModulesRepo,
         # don't try to update those that don't
         i = 0
         while i < len(modules_info):
             repo, module, _ = modules_info[i]
-            if repo.module_exists(module):
-                i += 1
-            else:
+            if not repo.module_exists(module):
                 log.warning(f"Module '{module}' does not exist in '{repo.fullname}'. Skipping...")
                 modules_info.pop(i)
+            elif sha is not None and not repo.sha_exists_on_branch(sha):
```

Comment:
I wonder if this sha is waht it is supposed to be. pylint says it is potentially undefined and it is a leftover of the loop in line 425.